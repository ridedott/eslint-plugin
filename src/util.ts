/* eslint-disable new-cap */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
import { ESLintUtils } from '@typescript-eslint/experimental-utils';
import { parse as parsePath } from 'path';

/**
 * This is not an import statement because it will make TSC copy the
 * package.json to the dist folder.
 */
const { version } = require('../package.json');

if (typeof version !== 'string') {
  throw new Error('Version field in package.json is not a string');
}

const REPO_URL = 'https://github.com/ridedott/eslint-config';

export const createRule = ESLintUtils.RuleCreator((name: string): string => {
  const ruleName = parsePath(name).name;

  return `${REPO_URL}/blob/v${version}/documentation/rules/${ruleName}.md`;
});
