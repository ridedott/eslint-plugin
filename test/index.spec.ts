/* eslint-disable @typescript-eslint/no-require-imports */
import { exists } from 'fs';
import { resolve } from 'path';
import { promisify } from 'util';

import internalRules = require('../src');

const existsAsync = promisify(exists);
const ruleNames = Object.keys(internalRules.rules);
const numberOfRules = 1;

const allConfigRules = Object.values(internalRules.configs)
  .map(
    (config: {
      plugins: string[];
      rules: { [key: string]: string };
    }): string[] => Object.keys(config.rules),
  )
  .reduce(
    (previousValue: string[], currentValue: string[]): string[] => [
      ...previousValue,
      ...currentValue,
    ],
    [],
  );

describe('rules', (): void => {
  it.each(ruleNames)(
    'should have a corresponding documentation file for rule: %s',
    async (rule: string): Promise<void> => {
      expect.assertions(1);

      const documentPath = resolve(__dirname, '../docs/rules', `${rule}.md`);
      const isDocumented = await existsAsync(documentPath);

      expect(isDocumented).toStrictEqual(true);
    },
  );

  it('should have the correct amount of rules', (): void => {
    expect.assertions(1);

    const { length } = ruleNames;

    expect(length).toStrictEqual(numberOfRules);
  });

  it('should export configs that refer to actual rules', (): void => {
    expect.assertions(3);

    const recommendedConfigs = internalRules.configs;

    expect(recommendedConfigs).toMatchSnapshot();
    expect(Object.keys(recommendedConfigs)).toStrictEqual([
      'all',
      'recommended',
    ]);
    expect(Object.keys(recommendedConfigs.all.rules)).toHaveLength(
      ruleNames.length,
    );
  });

  it.each(allConfigRules)(
    '%s should be included and prefixed',
    (rule: string): void => {
      expect.assertions(2);

      const ruleNamePrefix = '@ridedott/';
      const ruleName = rule.slice(ruleNamePrefix.length);
      expect(rule.startsWith(ruleNamePrefix)).toBe(true);
      expect(ruleNames).toContain(ruleName);
    },
  );
});
