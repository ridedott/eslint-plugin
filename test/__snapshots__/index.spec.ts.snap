// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rules should export configs that refer to actual rules 1`] = `
Object {
  "all": Object {
    "plugins": Array [
      "@ridedott",
    ],
    "rules": Object {
      "@ridedott/no-single-line-comment-block": "error",
      "@ridedott/no-template-literals-without-expression": "error",
    },
  },
  "recommended": Object {
    "plugins": Array [
      "@ridedott",
    ],
    "rules": Object {
      "@ridedott/no-single-line-comment-block": "warn",
      "@ridedott/no-template-literals-without-expression": "warn",
    },
  },
}
`;
